### ISSUE TITLE
Dot-repeating operator-pending leap movements causes "deadlock"

### ISSUE THREAD

**Issue by bjornevik** [ggandor/leap.nvim#50]
Nvim version: NVIM v0.8.0-dev-1132-g37a71d1f2

init.lua
```lua
require("leap").set_default_keymaps()
-- require("flit").setup({})
```

Use `nvim -u init.lua` to load minimal config

```lua
1234
```
Do a movement like `dz12` and try 

Having the same problem with `flit.nvim` and `lightspeed.nvim`.

**Comment by ggandor** [comment]
Does this problem occur after undoing a dot-repeat? That is a known Vim issue, according to vim-sneak's docs: https://github.com/justinmk/vim-sneak/blob/94c2de47ab301d476a2baec9ffda07367046bec9/doc/sneak.txt#L200-L202

However, except for the above scenario, dot-repeat should work fine.

**Comment by bjornevik** [comment]
> Does this problem occur after undoing a dot-repeat?

If I run through `nvim --clean` I'm able to `dta` -> dot-repeat -> undo -> dot-repeat as expected. If I attempt to do the same with `leap/flit` enabled I only get as far as `dta` -> dot-repeat, before nvim seizes up. If that's what you mean.

Might still be related to the issue referred to in vim-sneak docs, tried completely uninstall nvim-surround since I thought it might intefere, but still having this problem with nvim-surround disabled.

**Comment by ggandor** [comment]
> looks like the dot-repeat enters d~@ý, at least it's displayed in the bottom right.

Yep. It is also interesting that this only happens after _one_ dot-repeat - that is, dot-repeat -> dot-repeat+ -> undo works).
Note that you can exit from this frozen state with `C-c`m, if necessary.

**Comment by ggandor** [comment]
> I only get as far as dta -> dot-repeat, before nvim seizes up

To be clear: you experience this problem _without_ undo?

**Comment by bjornevik** [comment]
> Yep. It is also interesting that this only happens after one dot-repeat - that is, dot-repeat -> dot-repeat+ -> undo works).

That isn't the case for me, I'm stuck in the state until I either kill neovim, or enter `C-c`.

> To be clear: you experience this problem without undoing the change first?

Yes.

**Comment by bjornevik** [comment]
Tested some more today, seems I did not read the README properly and never added the tpope/vim-repeat dependency 🙈 

It works now, though you still have the issue where dot-repeat -> undo -> dot-repeat still has the issue. I'll leave it up to you whether or not to close this issue.

**Comment by bjornevik** [comment]
Have you considered changing the dot-repeat implementation as described in this gist? https://gist.github.com/kylechui/a5c1258cd2d86755f97b10fc921315c3 
Was thinking of trying to implement it and opening a PR, but am not really familiar with fennel. Also don't want to repeat work if you've already considered it and decided against it.

**Comment by ggandor** [comment]
> Tested some more today, seems I did not read the README properly and never added the tpope/vim-repeat dependency

Well... :D

> Have you considered changing the dot-repeat implementation as described in [this gist](https://gist.github.com/kylechui/a5c1258cd2d86755f97b10fc921315c3)?
Was thinking of trying to implement it and opening a PR, but am not really familiar with fennel. Also don't want to repeat work if you've already considered it and decided against it.

A PR is absolutely welcome, if the result is not significantly more verbose than the current implementation with vim-repeat. I haven't looked into it deeply yet, still don't really understand what we would need to do.

**Comment by gegoune** [comment]
gitsigns recently implemented (not sure if based on that gist) its own support for dot-repeat (without pope's plugin) as well. Just FYI.

**Comment by bjornevik** [comment]
Looked into it a couple of afternoons this past week, but haven't had the time to write any code.

A lot of it became clearer when I found [this fantastic blog post](https://www.vikasraj.dev/blog/vim-dot-repeat) by numToStr

How I've understood it is that the general pattern is to have a function callback that sets `vim.o.operatorfunc` to itself and returns `g@` when it is called "without a motion" (aka. called by user), and otherwise does whichever action when it is called with a motion. When `g@` is called the user is popped into operator-pending mode, and when the user enters the motion `g@` will call whatever function is set in `vim.o.operatorfunc` with the motion as an argument. Dot-repeat is automatically then set to `g@[motion]` with the same motion the user did and will repeat as expected. You can also set `vim.o.operatorfunc` 

Which should be relatively straight forward, but my non-existent experience with fennel (and lisp in general) is making it a bit harder to read the code and figure out what leap actually does and how everything fits together. Unsure if it's just a matter of changing a couple of lines in `fn set-dot-repeat* []` or if there needs to be a separate callback function for handling the different operators, visual mode, etc. Or even if it would constitute a significantly more verbose solution.

**Comment by FelipeLema** [comment]
changing to `operatorfunc` instead of depending `repeat.vim` is hard because the model for using `operatorfunc` needs to save a state needed to repeat the motion outside the command doing the motion.

Alternatively, one can try to re-use the current model (handling registers and operators ourselves), but that would involve handling corner cases that repeat.vim already handles (would require re-writing lots of code).
